File: 2024/cpp/day8.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <map>
#include <set>
#include <string>
#include <vector>

using namespace std;

int part1();
int part2();

int main() {
  cout << part1() << endl;
  cout << part2() << endl;
}

using Point = pair<int, int>;

auto get_two_antinodes(Point p1, Point p2) {
  int x1 = p1.first, y1 = p1.second;
  int x2 = p2.first, y2 = p2.second;
  int dx = x2 - x1;
  int dy = y2 - y1;
  Point a1 = {x1 - dx, y1 - dy};
  Point a2 = {x2 + dx, y2 + dy};

  return pair<Point, Point>(a1, a2);
}

bool node_in_bound(Point p, int w, int h) {
  return p.first >= 0 && p.first < h && p.second >= 0 && p.second < w;
}

int part1() {
  string line;
  ifstream input("../data/input8.txt");

  vector<string> grid;

  if (input.is_open()) {
    while (getline(input, line)) {
      grid.push_back(line);
    }
    input.close();
  }

  int H = grid.size();
  int W = grid[0].size();

  map<char, vector<Point>> antennas;

  for (int i = 0; i < H; i++) {
    for (int j = 0; j < W; j++) {
      if (grid[i][j] == '.')
        continue;
      antennas[grid[i][j]].push_back({i, j});
    }
  }

  set<Point> antinodes_at;

  for (auto keyval : antennas) {
    vector<Point> points = keyval.second;

    vector<pair<Point, Point>> pairs;

    for (size_t i = 0; i < points.size(); i++) {
      for (size_t j = 0; j < points.size(); j++) {
        if (i == j)
          continue;

        pairs.push_back({points[i], points[j]});
      }
    }

    set<pair<Point, Point>> visited_pairs;

    for (auto p : pairs) {
      Point a = p.first;
      Point b = p.second;

      pair<Point, Point> pair1 = {a, b};
      pair<Point, Point> pair2 = {b, a};
      if (find(visited_pairs.begin(), visited_pairs.end(), pair1) !=
          visited_pairs.end()) {
        continue;
      }
      if (find(visited_pairs.begin(), visited_pairs.end(), pair2) !=
          visited_pairs.end()) {
        continue;
      }
      visited_pairs.insert(pair1);

      auto antinodes = get_two_antinodes(a, b);
      Point a1 = antinodes.first, a2 = antinodes.second;

      if (node_in_bound(a1, H, W)) {
        antinodes_at.insert(a1);
      }
      if (node_in_bound(a2, H, W)) {
        antinodes_at.insert(a2);
      }
    }
  }

  return antinodes_at.size();
}

int part2() {
  string line;
  ifstream input("../data/input8.txt");

  vector<string> grid;

  if (input.is_open()) {
    while (getline(input, line)) {
      grid.push_back(line);
    }
    input.close();
  }

  int H = grid.size();
  int W = grid[0].size();

  map<char, vector<Point>> antennas;

  for (int i = 0; i < H; i++) {
    for (int j = 0; j < W; j++) {
      if (grid[i][j] == '.')
        continue;
      antennas[grid[i][j]].push_back({i, j});
    }
  }

  set<Point> antinodes_at;

  for (auto keyval : antennas) {
    vector<Point> points = keyval.second;

    vector<pair<Point, Point>> pairs;

    for (size_t i = 0; i < points.size(); i++) {
      for (size_t j = 0; j < points.size(); j++) {
        if (i == j)
          continue;

        pairs.push_back({points[i], points[j]});
      }
    }

    set<pair<Point, Point>> visited_pairs;

    for (auto p : pairs) {
      Point a = p.first;
      Point b = p.second;

      pair<Point, Point> pair1 = {a, b};
      pair<Point, Point> pair2 = {b, a};
      if (find(visited_pairs.begin(), visited_pairs.end(), pair1) !=
          visited_pairs.end()) {
        continue;
      }
      if (find(visited_pairs.begin(), visited_pairs.end(), pair2) !=
          visited_pairs.end()) {
        continue;
      }
      visited_pairs.insert(pair1);

      auto antinodes = get_two_antinodes(a, b);
      Point a1 = antinodes.first, a2 = antinodes.second;

      while (node_in_bound(a1, H, W) || node_in_bound(a2, H, W) ||
             node_in_bound(a, H, W) || node_in_bound(b, H, W)) {
        if (node_in_bound(a1, H, W)) {
          antinodes_at.insert(a1);
        }
        if (node_in_bound(a2, H, W)) {
          antinodes_at.insert(a2);
        }
        if (node_in_bound(a, H, W)) {
          antinodes_at.insert(a);
        }
        if (node_in_bound(b, H, W)) {
          antinodes_at.insert(b);
        }

        Point new_a1 = get_two_antinodes(a1, a).first;
        a = a1;
        a1 = new_a1;

        Point new_a2 = get_two_antinodes(b, a2).second;
        b = a2;
        a2 = new_a2;
      }
    }
  }

  return antinodes_at.size();
}

File: 2024/rust/src/day8.rs
use std::collections::{BTreeMap, BTreeSet};

use crate::day::{Day, Solution};

pub struct Day8 {
    day: Day,
}

impl Day8 {
    pub fn new() -> Self {
        Self { day: Day::new(8) }
    }
}

fn get_two_antinodes(p1: &(u32, u32), p2: &(u32, u32)) -> ((u32, u32), (u32, u32)) {
    let (x1, y1) = p1;
    let (x2, y2) = p2;
    let dx = x2 - x1;
    let dy = y2 - y1;
    let a1 = (x1 - dx, y1 - dy);
    let a2 = (x2 + dx, y2 + dy);

    (a1, a2)
}

impl Solution for Day8 {
    fn part1(&self) -> u64 {
        let input = include_str!("../../data/input8.txt");

        let mut antennas: BTreeMap<char, Vec<(u32, u32)>> = BTreeMap::default();

        let h = input.lines().count() as u32;
        let w = input.lines().next().unwrap().len() as u32;

        for (i, line) in input.lines().enumerate() {
            for (j, c) in line.chars().enumerate() {
                if c == '.' {
                    continue;
                }
                antennas
                    .entry(c)
                    .and_modify(|points| points.push((i as u32, j as u32)))
                    .or_insert(vec![(i as u32, j as u32)]);
            }
        }

        let mut antinodes: BTreeSet<(u32, u32)> = BTreeSet::default();

        for (_antenna, points) in antennas.iter() {
            let mut pairs: BTreeSet<((u32, u32), (u32, u32))> = BTreeSet::default();

            for i in 0..points.len() {
                for j in i + 1..points.len() {
                    if i == j {
                        continue;
                    }
                    pairs.insert((points[i], points[j]));
                }
            }

            for (a, b) in pairs.iter() {
                let (a1, a2) = get_two_antinodes(a, b);

                if a1.0 < h && a1.1 < w {
                    antinodes.insert(a1);
                }
                if a2.0 < h && a2.1 < w {
                    antinodes.insert(a2);
                }
            }
        }

        antinodes.len() as u64
    }

    fn part2(&self) -> u64 {
        let input = include_str!("../../data/input8.txt");

        let mut antennas: BTreeMap<char, Vec<(u32, u32)>> = BTreeMap::default();

        let h = input.lines().count() as u32;
        let w = input.lines().next().unwrap().len() as u32;

        for (i, line) in input.lines().enumerate() {
            for (j, c) in line.chars().enumerate() {
                if c == '.' {
                    continue;
                }
                antennas
                    .entry(c)
                    .and_modify(|points| points.push((i as u32, j as u32)))
                    .or_insert(vec![(i as u32, j as u32)]);
            }
        }

        let mut antinodes: BTreeSet<(u32, u32)> = BTreeSet::default();

        for (_antenna, points) in antennas.iter() {
            let mut pairs: BTreeSet<((u32, u32), (u32, u32))> = BTreeSet::default();

            for i in 0..points.len() {
                for j in i + 1..points.len() {
                    if i == j {
                        continue;
                    }
                    pairs.insert((points[i], points[j]));
                }
            }

            for (a, b) in pairs.into_iter() {
                let mut a = a.clone();
                let mut b = b.clone();

                let (mut a1, mut a2) = get_two_antinodes(&a, &b).clone();

                while [a1, a2, a, b].iter().any(|x| x.0 < h && x.1 < w) {
                    if a1.0 < h && a1.1 < w {
                        antinodes.insert(a1);
                    }
                    if a2.0 < h && a2.1 < w {
                        antinodes.insert(a2);
                    }
                    if a.0 < h && a.1 < w {
                        antinodes.insert(a);
                    }
                    if b.0 < h && b.1 < w {
                        antinodes.insert(b);
                    }

                    let new_a1 = get_two_antinodes(&a1, &a).0;
                    a = a1;
                    a1 = new_a1;
                    let new_a2 = get_two_antinodes(&b, &a2).1;
                    b = a2;
                    a2 = new_a2;
                }
            }
        }

        antinodes.len() as u64
    }

    fn get_day(&self) -> u8 {
        self.day.day
    }
}

pub fn run() {
    Day8::new().solve();
}
